### PR TITLE
Add toast bar

### PR DESCRIPTION
**Pull request by s-fletcher**
Screen Shot 2021-03-03 at 9 18 57 PM
Screen Shot 2021-03-03 at 9 19 13 PM

**Current behavior:** A toast can only be displayed one at a time. If `displayToast()` is called while a toast is displayed, it will ignore the request.

[ch20648](https://app.clubhouse.io/tailwind/story/20648/mobile-toast-bar)

<details>
<summary>File System Diff</summary>

```diff
@@ -130,12 +130,22 @@
   |  |  |  |-theme.ts
   |  |  |-PasswordInput
   |  |  |  |-index.tsx
   |  |  |-Button
   |  |  |  |-index.tsx
   |  |  |  |-theme.ts
+  |  |  |-Toast
+  |  |  |  |-Toast.tsx
+  |  |  |  |-ToastRenderer.tsx
+  |  |  |  |-index.tsx
+  |  |  |  |-theme.ts
+  |  |  |  |-useToast.ts
+  |  |  |  |-toastContext.ts
+  |  |  |  |-ToastProvider.tsx
+  |  |  |-Icon
+  |  |  |  |-index.tsx
   |-.buckconfig
   |-package.json
   |-README.md
   |-.github
   |  |-workflows
   |  |  |-build.yml
```
</details>

### CODE REVIEW

File: src/components/Icon/index.tsx [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+    source: Function;
+  }
+  | {
+    component: React.FC<SvgProps>;
+  };
+
+const iconMap: {
+  [index: string]: IconConfig;
+} = {
+  bouy: {
```
Comment (garrettjoecox) [inline src/components/Icon/index.tsx]: Let's remove all of these that came from triforce and just leave the one used in the toast

File: src/components/Toast/Toast.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import { StyleSheet, View } from 'react-native';
+import { useSafeAreaInsets } from 'react-native-safe-area-context';
+import { Button } from 'src/components/Button';
+import { Icon } from 'src/components/Icon';
+import { Text } from 'src/components/Text';
+import { ToastProps } from 'src/components/Toast/types';
+import { useToast } from 'src/components/Toast/useToast';
+import * as foundations from 'src/theme/foundations';
+
+const styles = StyleSheet.create({
```
Comment (garrettjoecox) [inline src/components/Toast/Toast.tsx]: Lets pair on converting this to the theming system when we have a chance

File: src/components/Toast/index.tsx [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export * from './Toast';
```
Comment (garrettjoecox) [inline src/components/Toast/index.tsx:1]: 👍

File: src/domains/home/HomeScreen.tsx [hunk 4]
```diff
@@ -1,9 +1,28 @@
 import React, { FC } from 'react';
+import { Button } from 'src/components/Button';
 import { Screen } from 'src/components/Screen';
 import { Text } from 'src/components/Text';
+import { useToast } from 'src/components/Toast/useToast';
```
Comment (garrettjoecox) [inline src/domains/home/HomeScreen.tsx]:
```suggestion
import { useToast } from 'src/components/Toast';
```

**Review by garrettjoecox — COMMENTED** [review]
Looks great! Good job!

For react-native-vector-icons how did you install? The installation seems a bit off, if it's not too much trouble I would recommend the [Option: with CocoaPods](https://github.com/oblador/react-native-vector-icons#option-with-cocoapods) for iOS and [Option: With Gradle](https://github.com/oblador/react-native-vector-icons#option-with-gradle-recommended) for android

File: __tests__/unit/init.js [hunk 5]
```diff
@@ -7,5 +7,5 @@ jest.mock('@react-native-async-storage/async-storage', () =>
 
 jest.mock('react-native-reanimated', () => require('react-native-reanimated/mock'));
 jest.mock('react-native-gesture-handler', () =>
-  require('react-native-gesture-handler/dist/src/__mocks__/RNGestureHandlerModule'),
+  require('react-native-gesture-handler/jestSetup'),
```
Comment (s-fletcher) [inline __tests__/unit/init.js:10]: I want to pull attention to this diff.

After pulling from `main` and running `npm install`, the path `react-native-gesture-handler/dist/src/__mocks__/RNGestureHandlerModule` no longer existed. 

Changing it to `react-native-gesture-handler/jestSetup` fixed the tests, but I'm not entirely sure the implications of this change.

File: src/components/Toast/theme.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
+import { Themeable } from '@emotion/react';
+import { ToastParts, ToastProps } from 'src/components/Toast';
+
+export const toastTheme: Themeable<ToastProps, ToastParts> = {
+  baseStyle: {
```
Comment (garrettjoecox) [inline src/components/Toast/theme.ts:5]: Everything work here as expected?

File: src/components/Toast/ToastProvider.tsx [hunk 7]
```diff
@@ -3,12 +3,16 @@ import React, { FC, useState } from 'react';
...
 export const ToastProvider: FC = ({ children }) => {
-  const [toast, setToast] = useState<ToastProps>();
+  const [toast, setToast] = useState<ToastOptions>();
   const DEFAULT_DURATION = 5000;
 
-  const displayToast = (options: ToastProps) => {
+  const displayToast = (options: ToastOptions) => {
     if (!toast) {
       LayoutAnimation.configureNext(LayoutAnimation.Presets.easeInEaseOut);
       setToast(options);
```
Comment (garrettjoecox) [inline src/components/Toast/ToastProvider.tsx:18]: Has the issue with multiple subsequent toasts been resolved?

File: __tests__/unit/init.js (hunk 5 above)
Comment (garrettjoecox) [reply __tests__/unit/init.js:10]: If you check the podfile.lock it looks like `pod update` or similar was called and a few unrelated deps were updated (namely `RNGestureHandler`), you can either revert the unrelated changes, or stick with this, as long as the tests pass.

This begins an ongoing battle with fighting to keep the tests running in RN...

**Review by garrettjoecox — APPROVED** [review]

File: src/components/Toast/ToastProvider.tsx (hunk 7 above)
Comment (garrettjoecox) [reply src/components/Toast/ToastProvider.tsx:18]: Bump on this ^

File: src/components/Toast/theme.ts (hunk 6 above)
Comment (s-fletcher) [reply src/components/Toast/theme.ts:5]: Yes! it took a little bit longer than expected to get it set up, but I see why it's very beneficial.

File: src/components/Toast/ToastProvider.tsx (hunk 7 above)
Comment (s-fletcher) [reply src/components/Toast/ToastProvider.tsx:18]: 🤦‍♂️ not sure how I missed this.

What did we agree on here? Do we want a queue system where toasts appear right after eachother? Or maybe subsequent toasts to appear below each other? I can add this feature in another PR.

Comment (garrettjoecox) [reply src/components/Toast/ToastProvider.tsx:18]: I don't have a strong opinion between those and just leaving it how it is now, as long as the timing issue was resolved. If I understood it correctly, an earlier implementation of this had the following problem:
- Submit 3000ms toast (#1)
- Wait 2500ms
- Submit another 3000ms toast (#2)
- The #2 toast would disappear after only 500ms, inheriting the old delay.

If this has been fixed you can disregard this and we can leave it as is.

Comment (s-fletcher) [reply src/components/Toast/ToastProvider.tsx:18]: Ahhh, yes. This has been fixed.

Comment (garrettjoecox) [reply src/components/Toast/ToastProvider.tsx:18]: Perfect 👍

**Pull request merged**